### ISSUE TITLE
initSession behavior onPlatformResume and onPlatformReady

### ISSUE THREAD

**Issue by FlorianWetschoreck** [BranchMetrics/cordova-ionic-phonegap-branch-deep-linking#412]
Hey,

I used the following code snippet which is called both on platform.ready and in platform.resume.subscribe

```
    Branch.initSession((data) => {
      alert('initSession Deep Link Data: ' + JSON.stringify(data));
    }).then((res) => {
      alert('Response: ' + JSON.stringify(res))
    }).catch((err) => {
      alert('Error: ' + JSON.stringify(err))
    });
```

and I observed the following behavior using ionic 3:

1) When the app starts:
- a: without clicked deeplink: only Response is called
- b: with clicked deeplink: BOTH initSession AND Response are called
2) When the app resumes:
- a: **without clicked deeplink: NOTHING is called**
- b: with clicked deeplink: initSession is called but not Response

My first question is, if this is the intended behavior?

**My most pressing question is:**
If this is the intended behavior: why is nothing called in case 2a? I want to handle this case in my app

A minor question is:
If it is the intended behavior: why are 1b and 2b inconsistent?

I am looking forward to your response :)

**Comment by ethanneff** [comment]
Hello,

Thank you for reaching out.

I apologize for the confusion. The `then` callback is only used for debugging purposes (to tell if the Branch.initSession() has been successfully initialized - which changes based on if Branch has been cached by the device or not). Ideally, you should only be listening for a response in the `(data)` callback.

Before:
```js
    Branch.initSession((data) => {
      alert('initSession Deep Link Data: ' + JSON.stringify(data));
    }).then((res) => {
      alert('Response: ' + JSON.stringify(res))
    }).catch((err) => {
      alert('Error: ' + JSON.stringify(err))
    });
```

After:
```js
    Branch.initSession((data) => {
      alert('initSession Deep Link Data: ' + JSON.stringify(data));
    });
```

Hope this clarifies everything.

**Comment by FlorianWetschoreck** [comment]
Hi,

thank you for your fast response.

Before I added the then() and catch() callbacks, I already tried your proposed solution.

However, if I observed this correctly, the initSession callback only gets called if the link is clicked and it does not get called when no link was clicked before opening the app.
Is this intended or is this unintended behavior?

I need to explicitly handle both (or rather all four) cases, so therefore the version with the additional callbacks at least let me explicitly handle 3 out of 4 cases.

I would be glad to hear your opinion on how to solve this issue.

**Comment by ethanneff** [comment]
Hello,

Yes, this is the intended behavior. The `Branch.initSession()` will only be called if data is passing through into the app. This happens whenever a user clicked on a link before opening the app.

Unfortunately, Branch does not have a way to handle **without clicked deeplink: NOTHING is called**. Typically, our partners have a default behavior for this edge case (e.g. default the home screen). If `Branch.initSession()` gets called, then they read the deep link data and navigate according (e.g. sign the user up or show a welcome screen).

I apologize for this inconvenience. Hopefully, this information helps.

**Comment by creativeround** [comment]
@ethanneff  you mentioned, 
"Branch does not have a way to handle **without clicked deeplink**: NOTHING is called."

Is this the same for a javascript that updates the 'window.location' to a deeplink?  For example, inside a safari view controller?  The user isn't physically clicking a link, but the javascript is telling the browser to navigate/change to the deeplink url.

**Comment by codinronan** [comment]
@ethanneff If I can weigh in: The `.then().catch()` version is incredibly helpful for the edge cases that @FlorianWetschoreck mentions. 

In particular, when an app has different logged-in/logged-out startup behavior, such as taking you to a dashboard screen (my app). Take for example the following scenario:

1. Default case: User is logged out, opens app by tapping app's icon. Goes to app home screen.
2. User is logged in, opens app by tapping app icon. Goes to dashboard screen.
3. User is logged OUT, opens app by tapping Branch link. What happens? Do they go to the home screen, or the deep link? How do I know that a deep link WILL BE PROCESSED? That is impossible.
4. User is logged IN, opens app by tapping branch link. Do they go to the dashboard, or the deep link? Same problem.

In cases 3 and 4, the presence of the `.then()` makes it possible to answer that question: all I need to do is wait for the then() to resolve. If it does, and I didn't get any deep link data, I can take the user to the normal screen; otherwise I can deep navigate.

Without the `.then()` I would be forced to FIRST navigate to Home/Dashboard, and THEN navigate to the deeplink. A jarring, terrible user experience and one we have data that shows causes engagement loss. Retention improved when I discovered I could capture the `.then()`. There was also a race condition where sometimes the deeplink would trigger first, then the auth event, pulling the user away from the link they clicked. 

By using the `.then()`,  all I have to do is wait for both the login/logout even, AND the branch `.then()`, and figure out the right thing to do, navigating only once.

I would suggest giving that promise first-class support, OR, always triggering the data callback even without data. Some apps simply need to know if Branch is going to trigger or not in order to provide the best user experience.

**Comment by FlorianWetschoreck** [comment]
I would like to stress a small extension to the post of @codinronan 

While the .then() helps you in cases, when the app is closed, it does not help you when the app resumes (as I stated in my initial post under scenario 2a).

Thus, you can not find out in a deterministic way if there might be a link clicked or not.
The result is, that if you open special content, there will always be a delay until you actually know which content to show. (This delay is caused by the link evaluation by the branch SDK.)
In my tests, this delay ranged between 500-1500ms. However, I also tested in cases with bad Internet connection and received the special response after 5000-10000ms which alienated the user.

Summing up, it would help us developers to deliver a seamless experience if the branch SDK could handle all four cases (see my original post) in a deterministic way.

**Comment by ethanneff** [comment]
Hello everybody!

I apologize for the delay. I got pull into another team so the Branch Cordova SDK hasn't seen much love. The good news is I just updated the Branch Cordova SDK. I would recommend updating to the latest `2.7.0` version. You can use this [guide](https://github.com/BranchMetrics/cordova-ionic-phonegap-branch-deep-linking#compiling-updating-the-branch-sdk) to update the SDK. Keep in mind, this version has a breaking change. The following change will need to be done to your code to read the deep link data. You can find more infomation about reading deep link data [here](https://github.com/BranchMetrics/cordova-ionic-phonegap-branch-deep-linking#read-deep-link).

```diff
  // Branch initialization
- Branch.initSession(function(data) {
+ Branch.initSession().then(function(data) {
    if (data["+clicked_branch_link"]) {
      // read deep link data on click
      alert("Deep Link Data: " + JSON.stringify(data));
    }
  });
```

Thanks,

**Comment by codinronan** [comment]
@ethanneff that's a great change! Exactly what we needed. Thank you!! :) Will make sure to integrate that in my next native deploy.
